### ISSUE TITLE
Changing ippool to use ipip.mode=cross-subnet (from always) doesn't update all routes

### ISSUE THREAD

**Issue by bradbehle** [projectcalico/calico#1584]
I originally reported this problem with Calico 2.6.1 in this issue: https://github.com/projectcalico/calico/issues/1325    It was thought to be fixed in Calico 2.6.3, but I was able to recreate it with Calico 2.6.5.  Here is a description:

When I update calico-node to 2.6.5 and to use kube-hosted calico and IP_AUTODETECTION_METHOD=interface=eth2 (from systemd managed calico-node using hardcoded IP) on all worker nodes, I then update the ippool to set ipip.mode=cross-subnet (from its original setting of "always").  All the worker nodes are in the same subnet.  This results in some of the routes being updated to use "eth2" instead of tunl0 appropriately, but some routes remain using tunl0.  This results in pods on some nodes not being able to communicate with pods on other nodes (because one side routes to tunl0 and the other side routes to eth2).

## Expected Behavior
I expect that all the routes would be updated to use eth2

## Current Behavior
Only some of the routes are updated some times.  Other times when I try this all the routes are properly updated.

## Steps to Reproduce (for bugs)

1. Install k8s cluster with kube 1.5.6 and calico 2.1.5 (calico-node running as a systemd service, with hard coded IPs so calico sees them as X.X.X.X/32) using IP in IP encapsulation with the mode set to Always.  Create a cluster with at least three worker nodes.
2. Update the k8s cluster to kube 1.7.4 and calico 2.6.5 (calico-node running hosted by kubernetes, with IP_AUTODETECTION_METHOD=interface=eth2), and don't change the ippool yet.  Note that this modifies the calico node resources to use the same IP as before, but properly detects the subnet as /24.  At this point calico is properly still routing all pod to pod traffic through tunl0 since ip in ip mode is Always
3. Once all workers have been updated as mentioned above, change the ippool to set mode to "cross-subnet" using calicoctl
4. This causes some of the routes to get immediately updated to route through eth2 (and not tunl0), but some remain routing through tunl0.

Here are the routes from three nodes in the cluster I most recently recreated this from, including the nodes and ippool:

~~~
Node: 10.110.100.2 routing table:
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.0.2.2        0.0.0.0         UG    0      0        0 eth0
10.0.2.0        0.0.0.0         255.255.255.0   U     0      0        0 eth0
10.100.100.0    0.0.0.0         255.255.255.0   U     0      0        0 eth2
172.16.2.192    10.100.100.4    255.255.255.192 UG    0      0        0 eth2
172.16.78.64    10.100.100.3    255.255.255.192 UG    0      0        0 eth2
172.16.243.0    0.0.0.0         255.255.255.192 U     0      0        0 *
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
192.168.10.0    0.0.0.0         255.255.255.0   U     0      0        0 eth1

Node: 10.110.100.3 routing table:
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.0.2.2        0.0.0.0         UG    0      0        0 eth0
10.0.2.0        0.0.0.0         255.255.255.0   U     0      0        0 eth0
10.100.100.0    0.0.0.0         255.255.255.0   U     0      0        0 eth2
172.16.2.192    10.100.100.4    255.255.255.192 UG    0      0        0 eth2
172.16.78.64    0.0.0.0         255.255.255.192 U     0      0        0 *
172.16.78.67    0.0.0.0         255.255.255.255 UH    0      0        0 cali1939eb3f573
172.16.78.68    0.0.0.0         255.255.255.255 UH    0      0        0 cali6200adae3fd
172.16.78.69    0.0.0.0         255.255.255.255 UH    0      0        0 cali53ac757ec83
172.16.78.70    0.0.0.0         255.255.255.255 UH    0      0        0 cali924c5d044fe
172.16.78.71    0.0.0.0         255.255.255.255 UH    0      0        0 calib071657af4e
172.16.78.72    0.0.0.0         255.255.255.255 UH    0      0        0 calia3c56a24e0c
172.16.78.73    0.0.0.0         255.255.255.255 UH    0      0        0 calibf9f9964cf7
172.16.78.74    0.0.0.0         255.255.255.255 UH    0      0        0 cali1de74324c6e
172.16.243.0    10.100.100.2    255.255.255.192 UG    0      0        0 tunl0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
192.168.10.0    0.0.0.0         255.255.255.0   U     0      0        0 eth1

Node: 10.110.100.4 routing table:
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.0.2.2        0.0.0.0         UG    0      0        0 eth0
10.0.2.0        0.0.0.0         255.255.255.0   U     0      0        0 eth0
10.100.100.0    0.0.0.0         255.255.255.0   U     0      0        0 eth2
172.16.2.192    0.0.0.0         255.255.255.192 U     0      0        0 *
172.16.2.195    0.0.0.0         255.255.255.255 UH    0      0        0 cali82ecd7401af
172.16.2.196    0.0.0.0         255.255.255.255 UH    0      0        0 cali1af28459e88
172.16.2.197    0.0.0.0         255.255.255.255 UH    0      0        0 cali83caebbedf7
172.16.2.198    0.0.0.0         255.255.255.255 UH    0      0        0 cali7f47e5b2464
172.16.2.199    0.0.0.0         255.255.255.255 UH    0      0        0 cali33aa9e33f65
172.16.78.64    10.100.100.3    255.255.255.192 UG    0      0        0 eth2
172.16.243.0    10.100.100.2    255.255.255.192 UG    0      0        0 tunl0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
192.168.10.0    0.0.0.0         255.255.255.0   U     0      0        0 eth1

vagrant@carrier0-worker-1:~$ /opt/bin/calicoctl get ippool -o yaml
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: 172.16.0.0/16
  spec:
    ipip:
      enabled: true
      mode: cross-subnet
    nat-outgoing: true
vagrant@carrier0-worker-1:~$ /opt/bin/calicoctl get nodes -o wide
NAME                ASN       IPV4              IPV6   
carrier0-master-1   (64512)   10.100.100.2/24          
carrier0-worker-1   (64512)   10.100.100.3/24          
carrier0-worker-2   (64512)   10.100.100.4/24          

vagrant@carrier0-worker-1:~$ 
~~~

And here are the calico-node logs from two of the three nodes.  I didn't get the logs from 10.100.100.2 before it wrapped, and also, in this case 10.100.100.2 is still running managed by systemd, because it is not a k8s worker node it is the k8s master node.  I changed the ippool to use cross-subnet at around 2018-01-05 18:00:57

[calico-node-10.100.100.4.log](https://github.com/projectcalico/calico/files/1610612/calico-node-10.100.100.4.log)
[calico-node-10.100.100.3.log](https://github.com/projectcalico/calico/files/1610613/calico-node-10.100.100.3.log)

## Context

This results in pods on some nodes not being able to communicate with pods on other nodes (because one side routes to tunl0 and the other side routes to eth2).

## Your Environment

* Calico version: upgrading to 2.6.5 (have also seen this while updating to 2.6.1 as mentioned in the issue I linked to above, so I don't think it is something introduced recently).
* Orchestrator version (e.g. kubernetes, mesos, rkt): k8s 1.7.4
* Operating System and version: Ububtu 16.04
* Link to your project (optional):

**Comment by robbrockbank** [comment]
Hi @bradbehle :  The logs indicate that we're no longer hitting the issue that I previously saw where a transient issue with the config resulted in the config not being reloaded by bird.  I notice that for the .3 node the calico/node instance seems to have been restarted at about the time your made the config change, whereas the .4 node seems to do a standard config update without needing to restart the node.  I'm not sure of the significance of that (if any) - but just to document here.

To make sure I've got all the details correct, can you confirm the following is true:
-  Restarting the calico/node instances does not fix the problem.
-  Reconfiguring the IPPool (re-applying the same config) does not fix the problem
-  Deleting the route fixes the problem - the route gets re-instated correctly.

Do you happen to know if you have multiple broken routes, whether deleting a single route fixes all of them, or just the one you deleted?

**Comment by neiljerram** [comment]
@bradbehle I'm unclear whether you think this is a regression since 2.6.3, or that it wasn't actually fixed in 2.6.3. Please could you clarify that?

**Comment by bradbehle** [comment]
I have seen this with both Calico 2.6.1 and calico 2.6.5, so no, I do not think this is a regression.  I opened issue #1325 when I saw it in Calico 2.6.1.  I did some debugging on a cluster in this state yesterday with Grant, hopefully he can get you those logs and output if he hasn't already

**Comment by bradbehle** [comment]
@robbrockbank 
1) Restarting calico-node doesn't completely fix the problem.  When I ran it just now, it did fix up one of the routes, but the others were not fixed.
2) Reapplying the ippool has no effect
3) Most clusters do have multiple bad routes (routes that go through tunl0 when they shouldn't), and each bad route needs to be removed before it is replaced with the new one.  Deleting one bad route does NOT fix all of them up

I've attached the output from running the above tests so you can see for yourself, and maybe catch something I didn't (I changed the IPs and hostnames, but everything else is exactly as it happened).  As I mentioned above, I did some debugging on a bigger cluster (8 worker nodes, with many more bad routes) with Grant yesterday, and got some bird debug logs, output from bird commands, etc.  He has that output and those logs if those might be helpful.
[debug_output_1_17_2018.txt](https://github.com/projectcalico/calico/files/1639381/debug_output_1_17_2018.txt)

**Comment by robbrockbank** [comment]
Hi @bradbehle :  I've just merged a couple of PRs that should address the problem in the 2.6.7 release:

#1625 and https://github.com/projectcalico/bird/pull/53

The diags you provided were very useful in helping me spot that there was a specific code path where the routes *may* not get updated during a route scan.

**Comment by caseydavenport** [comment]
Fixed by #1625 - hopefully releasing in v2.6.7 later today.
